tests/test_gamecontroller.py

```diff
@@ -1,137 +1,70 @@
-        mock_deck.__len__.return_value = 0
-        mock_player = create_autospec(Player)
-        mock_game_state = create_autospec(GameState)
-        mock_board = create_autospec(Board)
-        gc = GameController([mock_player, mock_player])
-        gc.deck = mock_deck
-        gc.master_game_state = mock_game_state
-        gc.master_game_state.board = mock_board
-        gc.master_game_state.board.fuse_tokens = 1
-        gc.master_game_state.board.game_almost_over = 0
-        self.assertIs(gc.game_over(0, gc.master_game_state), True)
-
-    def test_game_over_not_over_deck_not_empty(self):
-        mock_deck = create_autospec(Deck)
-        mock_deck.__len__.return_value = 1
-        mock_player = create_autospec(Player)
-        mock_game_state = create_autospec(GameState)
-        mock_board = create_autospec(Board)
-        gc = GameController([mock_player, mock_player])
-        gc.deck = mock_deck
-        gc.master_game_state = mock_game_state
-        gc.master_game_state.board = mock_board
-        gc.master_game_state.board.fuse_tokens = 1
-        gc.master_game_state.board.game_almost_over = 0
-
-        self.assertIs(gc.game_over(0, gc.master_game_state), False)
-
-    def test_game_over_not_over_deck_empty(self):
-        mock_deck = create_autospec(Deck)
-        mock_deck.__len__.return_value = 0
-        mock_player = create_autospec(Player)
-        mock_game_state = create_autospec(GameState)
-        mock_board = create_autospec(Board)
-        gc = GameController([mock_player, mock_player])
-        gc.deck = mock_deck
-        gc.master_game_state = mock_game_state
-        gc.master_game_state.board = mock_board
-        gc.master_game_state.board.fuse_tokens = 1
-        gc.master_game_state.board.game_almost_over = 0
-
-        self.assertIs(gc.game_over(1, gc.master_game_state), False)
-
-
-    # TODO: Test play_game fails if player cannot make_move (may belong in make_move?)
-    def test_play_game_last_move_wins(self):
-        # Because if the final move causes a win we should return true but the game should end with correct points
-        pass
+@pytest.fixture(scope='module')
+def mock_player():
+    return create_autospec(Player)
+
+
+@pytest.fixture(scope="module")
+def mock_deck():
+    return util.create_mock_deck(r_draw_card='card')
+
+
+@pytest.fixture(scope='function')
+def gc(mock_player):
+    _gc = GameController([mock_player, mock_player])
+    _gc.master_game_state = util.create_mock_gamestate()
+    return _gc
+
+
+@pytest.mark.parametrize(argnames='player_count', argvalues=[1, 6])
+def test_constructor_fail_invalid_number_of_players(player_count, mock_player):
+    with pytest.raises(ValueError) as excinfo:
+        GameController(players=[mock_player for _ in range(player_count)])
+    assert str(excinfo.value) == 'There must be between 2 and 5 players to play Hanabi.'
+
+
+@pytest.mark.parametrize(argnames=['player_count', 'expected_hand_size'], argvalues=[[2, 5], [3, 5], [4, 4], [5, 4]])
```

Comment:
Slightly confusing, but maybe that's just because I'm not familiar with parameterize.

Reply:
Definitely, looking for feedback on these. Parametrize is nifty and powerful but can lower readability.   In this case, the test checks that `GameController` is dealing the correct amount of cards for games with 2, 3, 4, and 5 players. Each sublist (e.g. [2, 5] or [4, 4]) contains 2 values: the number of players (`player_count`) and the number of cards a player's hand should contain after dealing (`expected_hand_size`), respectively.